### DISCUSSION TITLE
Binding viewmodel to xaml, what am I missing?

### DISCUSSION THREAD

**Discussion by sej69** [dotnet/maui#11084] [Q&A] [answered]
I'm just learning this toolkit and I'm struggling to understand why my binding is reporting the viewmodel as type string.  I'm trying to use an ObservableCollection of objects.

this is my ViewModel:

`
    public partial class MainViewModel : ObservableObject
    {
    private string APIKey = "";

    public MainViewModel()
    {
        Items = new ObservableCollection<Customer>();
        AdminAPI adminAPI = new AdminAPI(APIKey);

        Task.Run( () => LoadData());

    }

    [ObservableProperty]
    ObservableCollection<Customer> items;

    [ObservableProperty]
    string text;

    [RelayCommand]
    public async Task LoadData()
    {
        AdminAPI adminAPI = new AdminAPI(APIKey);
        Items = await adminAPI.GetCustomersAsync();
    }

    [RelayCommand]
    void Add()
    {
        if (string.IsNullOrWhiteSpace(Text))
            return;

        Customer customer = new Customer();
        customer.CustomerName = Text;

        Items.Add(customer);
        Text = string.Empty;
    }

    [RelayCommand]
    async Task Tap(string s)
    {
        await Shell.Current.GoToAsync($"{nameof(CustomerDetailPage)}?Text={s}");
    }

    }
`

This is my XAML:

`
    <ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="TestApp.MainPage"
             xmlns:viewmodel="clr-namespace:TestApp.ViewModels"
             x:DataType="viewmodel:MainViewModel"
             >

    <Grid RowDefinitions="100, Auto, *"
          ColumnDefinitions=".50*, .25*, .25*"
          Padding="10">

        <Image Grid.ColumnSpan="3"
               Source="logo.png"
               BackgroundColor="Transparent"/>

        <Entry Placeholder="Enter Text" Grid.Row="1" Text="{Binding Text}" />
        <Button Text="Search" Grid.Row="1" Grid.Column="1" />
        <Button Text="Add" Grid.Row="1" Grid.Column="2"  Command="{Binding AddCommand}"/>
        
        <CollectionView Grid.Row="2" Grid.ColumnSpan="3" ItemsSource="{Binding Items}" SelectionMode="None">
            <CollectionView.ItemTemplate>
                <DataTemplate x:DataType="x:Type x:String">
                    <SwipeView>
                        <SwipeView.RightItems>
                            <SwipeItem Text="Delete" BackgroundColor="Red"/>
                        </SwipeView.RightItems>
                        <Grid Padding="0,5">
                            <Frame>
                                <Frame.GestureRecognizers>
                                    <TapGestureRecognizer 
                                        Command="{Binding Source={RelativeSource AncestorType={x:Type viewmodel:MainViewModel}}, Path=TapCommand}"
                                        CommandParameter="{Binding .}" />
                                </Frame.GestureRecognizers>
                                <Label Text="{Binding .}" FontSize="20"/>
                            </Frame>
                        </Grid>
                    </SwipeView>
                </DataTemplate>
            </CollectionView.ItemTemplate>
        </CollectionView>
    </Grid>

    </ContentPage>
`

When I hover over the viewModel:MainViewModel above it says type string.

What I want to do is bind the label with the "CustomerName" instead of {Binding .}  So CommandParameter = {Binding CustomerName}

But when I do that, I get a compile error: "CustomerName not found on "System.String"

CollectionView:  <CollectionView Grid.Row="2" Grid.ColumnSpan="3" ItemsSource="{Binding Items}" SelectionMode="None">

is binding to Items which defined by ObservableCollection<Customer> items in the above viewModel.

any idea what I could be doing wrong?

**Comment by sjordanGSS** [comment]
You're upsetting the compiled binding system because your datatemplate thinks that the Items list contains strings
```
<DataTemplate x:DataType="x:Type x:String">
```

Change the value of x:DataType to point to your Customer model and you should be able to bind to its properties within the datatemplate.

**Comment by mhrastegari** [comment]
you should add this too in your xaml.
`
<ContentPage.BindingContext>
<vm:MainViewModel />
</ContentPage.BindingContext>
`

**Comment by sej69** [comment]
Thanks for the suggestions.

I am still missing something.  I did not modify the viewModel from the above code.

This is my current XAML:

    <?xml version="1.0" encoding="utf-8" ?>
    <ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="LicenseApp.MainPage"
             xmlns:viewmodel="clr-namespace:LicenseApp.ViewModels"
             x:DataType="viewmodel:MainViewModel"
             >

    <ContentPage.BindingContext>
        <viewmodel:MainViewModel />
    </ContentPage.BindingContext>
    
    <Grid RowDefinitions="100, Auto, *"
          ColumnDefinitions=".50*, .25*, .25*"
          Padding="10">

        <Image Grid.ColumnSpan="3"
               Source="logo.png"
               BackgroundColor="Transparent"/>

        <Entry Placeholder="Enter Text" Grid.Row="1" Text="{Binding Text}" />
        <Button Text="Search" Grid.Row="1" Grid.Column="1" />
        <Button Text="Add" Grid.Row="1" Grid.Column="2"  Command="{Binding AddCommand}"/>
        
        <CollectionView Grid.Row="2" Grid.ColumnSpan="3" ItemsSource="{Binding Items}" SelectionMode="None">
            <CollectionView.ItemTemplate>
                <DataTemplate x:DataType="viewmodel:MainViewModel">
                    <SwipeView>
                        <SwipeView.RightItems>
                            <SwipeItem Text="Delete" BackgroundColor="Red"/>
                        </SwipeView.RightItems>
                        <Grid Padding="0,5">
                            <Frame>
                                <Frame.GestureRecognizers>
                                    <TapGestureRecognizer 
                                        Command="{Binding Source={RelativeSource AncestorType={x:Type viewmodel:MainViewModel}}, Path=TapCommand}"
                                        CommandParameter="{Binding .}" />
                                </Frame.GestureRecognizers>
                                <Label Text="{Binding CustomerName}" FontSize="20"/>
                            </Frame>
                        </Grid>
                    </SwipeView>
                </DataTemplate>
            </CollectionView.ItemTemplate>
        </CollectionView>
    </Grid>

    </ContentPage>

The line :< Label Text="{Binding CustomerName}" FontSize="20"/ > won't resolve the "CustomerName" in Items.Customer name.  I've also tried Items.Customer name there.  I'm hoping it's something simple?

**Comment by mhrastegari** [comment, marked as answer]
In `x:DataType` as it says you should exactly specify what TYPE of data you wanna show the user.
In almost most of the time you should give it your `Model`.

if we think that items in `ItemsSource="{Binding Items}` is gives us items from the `Customers` class which has some properties like `Name`, so you should write it like this:
```
<DataTemplate x:DataType="model:Customers">
                <SwipeView>
                    <SwipeView.RightItems>
                        <SwipeItem Text="Delete" BackgroundColor="Red"/>
                    </SwipeView.RightItems>
                    <Grid Padding="0,5">
                        <Frame>
                            <Frame.GestureRecognizers>
                                <TapGestureRecognizer 
                                    Command="{Binding Source={RelativeSource AncestorType={x:Type viewmodel:MainViewModel}}, Path=TapCommand}"
                                    CommandParameter="{Binding .}" />
                            </Frame.GestureRecognizers>
                            <Label Text="{Binding Name}" FontSize="20"/>
                        </Frame>
                    </Grid>
                </SwipeView>
            </DataTemplate>
```

**Comment by sej69** [comment]
Ah got it. Thanks.  

I have an external library that I intend on linking with which contains all of the shareable classes between the azure function and the .net Maui App.

If I move the class to the .net Maui App, and I provide an xlmns: config in the xaml, the program compiles and runs.  But I can't seem to figure out how to link in an external library. I'll poke around with that some more...

Thanks for your help!

**Comment by mhrastegari** [comment]
Happy to help.
